reps/2023-10-30-gpu-memory-support.md

```diff
@@ -0,0 +1,274 @@
+
+## Stewardship
+
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@pcmoritz, @jjyao, @scv119
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@jjyao
+
+## Design
+
+### API
+Users will be able to specify amount of gpu memory to their ray tasks/actors using `gpu_memory` on `ray.remote`. The specified `gpu_memory` will be the amount of gpu resources from a single gpu that will be allocated to users ray tasks/actors.
+
+```python
+# Request a fractional GPU with specified gpu_memory in bytes.
+# Mutually exclusive with num_gpus.
+@ray.remote(gpu_memory=1024 * 1024 * 1024) # 1 mb request
+def task:
+  …
+```
+
+When a Ray node is started, Ray will auto detect the number of GPUs and GPU memory of each GPU and set the GPU and gpu_memory resources accordingly. Users also have the option to manually specify `gpu_memory` resources as the sum of total gpu memory across all gpus in the node. The default value is `num_gpus` * total memory of the gpu type in the node.
+
+```bash
+ray start # auto detection
+
+ray start –num_gpus=3 –gpu_memory=3000 * 1024 * 1024 * 1024 # manual override, each gpu has 1000mb total memory
+```
+
+Note that GPU memory and compute unit is 1-1 conversion, means 20GB of gpu memory is equivalent to 0.5 fractional value of an `A100_40GB` gpu. So, for simplicity and consistency, ray doesn't allow users to specify both `num_gpus` and `gpu_memory` in a single ray task/actor.
+
+```python
+# Request a fractional GPU both num_gpus and gpu_memory is not allowed
+@ray.remote(gpu_memory=1024 * 1024 * 1024, num_gpus=0.5) # raise ValueError exception
+def not_allowed_task:
+  …
+
+# Request a fractional GPU with specified num_gpus.
+@ray.remote(num_gpus=0.5)
+def num_gpus_task:
+  …
+
+# Request a fractional GPU with specified gpu_memory.
+@ray.remote(gpu_memory=1024 * 1024 * 1024) 
+def gpu_memory_task:
+  …
+```
+
+Additionally, users can still specify which gpu type they want to use by specifying `accelerator_type`.
+
+```python
+# Request a fractional of A100 GPU with specified gpu_memory
+@ray.remote(gpu_memory=1024 * 1024 * 1024 * 1024, accelerator_type="NVIDIA_A100")
+def nvidia_a100_gpu_task:
+  …
+
+# Requesting 30GB of gpu memory from a A10 GPU with 24GB of memory.
+# Task won't be able to be scheduled.
+@ray.remote(gpu_memory=30 * 1024 * 1024 * 1024 * 1024, accelerator_type="NVIDIA_TESLA_A10G")
+def nvidia_a10_gpu_task:
+  …
+```
+
+#### Placement Group
+TBD
```

Comment:
Can you fill this in?

It should be like `placement_group([{"GPU_MEMORY": 1024}, {CPU: 1}])`